Nikita Samak

SiegeOder opened the pull request.

SiegeOder committed:
project uploading

SiegeOder committed:
Update setup.py

tests excluded

SiegeOder committed:
update tests

importing problems

SiegeOder committed:
update tests

SiegeOder committed:
removed imports

AlexeiBuzuma commented on final_task/pycalc/constants_and_operations.py:
```diff
@@ -0,0 +1,25 @@
+from pycalc.module_appender import (append_module_by_name)
+
+operations = {'^': lambda arg1, arg2: arg1 ** arg2,
+              '/': lambda arg1, arg2: arg1 / arg2 if arg2 != 0 else exit("ERROR: division by zero"),
```
советую посмотреть в сторону модуля `operators`. Нет необходимости самому реализовывать операторы через лямбды, учитывая, что они уже есть в стандартной библиотеке

SiegeOder replied:
Подумал, может выгоднее написать самому, чем подгружать целый модуль. 
Исправил

AlexeiBuzuma commented on final_task/pycalc/constants_and_operations.py:
```diff
@@ -0,0 +1,25 @@
+                      '<': lambda arg1, arg2: arg1 < arg2,
+                      }
+
+constants = {}
```
Названия констнат должны быть аппер кейсом через нижнее подчеркивание
LOGICAL_OPERATIONS
FUNCTIONS 
и т.д.

SiegeOder replied:
пеп8, виноват. Исправил

AlexeiBuzuma commented on final_task/pycalc/expression_calculator.py:
```diff
@@ -0,0 +1,112 @@
+def prepare_expression(expression: list):
+    if not expression:
+        print(f"ERROR: empty expression")
+        exit(6)
```
для решения такого объема в коде должна быть только одна точка выхода, никак не больше одной. А их тут намного больше, чем одна. функции exit и исключения SystemExit скорее всего если и должны использоваться, то только в одном месте. 

+ почему в функцию exit отправляется 6?

SiegeOder replied:
Исправил, теперь для выхода используется raise Exception и функцию трейсбека выполняет она, а не цифра системного выхода)0))

AlexeiBuzuma replied:
На самом деле это не исправляет проблему. Суть именно в том, чтобы в коде было только одно место, где программа завершает свою работу. В случае с функциями exit таких точек в программе было много. То есть мы не можем с уверенностью сказать, где в коде завершающий кусок кода, потому что точки выхода разбросаны по всему приложению. 

Замена этой функции на исключение ничего не исправляет в этом случае. 

Советую сделать примерно следующую вещь: 
В файле `pycalc.py` внутри функции main сделать большой try except, который будет отлавливать все необработанные исключения в приложении и завершать приложение. Например так: 
```python
def main():
    try:
        func()
        foo()
    except Exception as e:
        print('ERROR: {}'.format(e)')
```

а в местах, где происходит ошибка, кидать примерно таким образом исключения: 
```python
raise Exception('Unbalanced brackets')
```

SiegeOder replied:
Спасибо, очень хороший совет и пример! Так и сделал.

AlexeiBuzuma commented on final_task/pycalc/valid_checks.py:
```diff
@@ -0,0 +1,31 @@
+        exit(2)
+
+    if search(r'^[-/*+<=>]+$',
+              expression) or search(r'[!=<>*/]+\s+[!=<>*/]',
```
какие-то странные отступы :)

SiegeOder replied:
согласен, исправил

SiegeOder committed:
update

- changed lamdas for built-in
- uppercased constants
- changed sys.exits for raises
- idents is not so strange now

AlexeiBuzuma commented on .gitignore:
```diff
@@ -102,3 +102,15 @@ venv.bak/
 
 # mypy
 .mypy_cache/
+.idea/misc.xml
```
эти изменения можно записать короче (исключить не конкретные файлы, а просто целую папку):
```
.idea/
final_task/.idea
```

SiegeOder replied:
Спасибо, исправил.

AlexeiBuzuma commented on final_task/pycalc/expression_calculator.py:
```diff
@@ -0,0 +1,112 @@
+
+def calculate_logical_expression(expression: list) -> bool:
+    for logical_operation in LOGICAL_OPERATIONS.keys():
+        if logical_operation in expression:
```
В случае, если это условие не будет выполнено, то функция вернет `None`, что не соответствует определению функции (судя определению должен вернуться тип bool)

SiegeOder replied:
Добавил исключение, если не выполняется условие. Проверку наличия оператора выполняет другая функция.

AlexeiBuzuma commented on final_task/pycalc/expression_calculator.py:
```diff
@@ -0,0 +1,112 @@
+            return LOGICAL_OPERATIONS[logical_operation](expression[0], expression[2])
+
+
+def calculate_expression(expression: list) -> list:
```
В одном случае функция вернет результат функции `calculate`, а в другой не возвращает ничего. 
Я бы посоветовал соблюдать единый интерфейс у функций.

SiegeOder replied:
Немного не понял, в какой ситуации функция calculate_expression не возвращает ничего?

AlexeiBuzuma commented on final_task/pycalc/pycalc.py:
```diff
@@ -0,0 +1,25 @@
+from pycalc.constants_and_operations import (CONSTANTS, FUNCTIONS)
+from pycalc.expression_calculator import (calculate_expression)
```
в данном случае скобочки не нужны

SiegeOder replied:
Лишние скобки в импортах из программы удалены.

SiegeOder committed:
Update

- exit point is only one now
- exceptions added to function to match definition
- unnecessary brackets removed
- updated gitignore

Aseedr commented on final_task/pycalc/expression_calculator.py:
```diff
@@ -0,0 +1,109 @@
+        if value in ['^', '//', '/', '%', '*', '+', ',']:
+            if expression[index + 1] == '-':
+                expression[index + 2] = -convert(expression[index + 2])
+                del expression[index + 1]
```
Во время итерации по объекту expression, идет удаление внутренних элементов, как следствие может возникнуть ошибка с индексацией.

dzhigailo commented on final_task/pycalc/constants_and_operations.py:
```diff
@@ -0,0 +1,27 @@
+
+from pycalc.module_appender import (append_module_by_name)
+
+OPERATIONS = {'^': operator.pow,
```
It's recommended to leave firt row empty in multiline statements
```suggestion
OPERATIONS = {
    '^': operator.pow,
```

dzhigailo commented on final_task/pycalc/expression_calculator.py:
```diff
@@ -0,0 +1,109 @@
+    return start_index, end_index
+
+
+def remove_brackets(expression: list):
```
Here and below:
If one defines function annotations, function output type should be defined as well

dzhigailo commented on final_task/pycalc/module_appender.py:
```diff
@@ -0,0 +1,22 @@
+    try:
+        module = __import__(module_name)
+        append_module(module, constants, functions)
+    except Exception as e:
```
`Exception` is too broad for this case. Except `ImportError` will be right.
